### ISSUE TITLE
Should Non-Nullable Assertions (x!.foo) conditionally throw?

### ISSUE THREAD

**Issue by robpalme** [tc39/proposal-type-annotations#31]
Non-Nullable Assertions using postfix `!` (e.g. `x!.foo`) are currently proposed to be erasable.  This issue is to track discussion on whether this operator in JavaScript deserves to have runtime effects, e.g. throwing if the operand is `undefined` or `null`.

It is unlikely this proposal would be expanded to include runtime behavior, so the consequence on this proposal is to decide if the operator is either:

- **in-scope** because the operator has no runtime effects - just like TypeScript today
...or...
- **out-of-scope:** because the operator should conditionally throw at runtime, as discussed [here](https://es.discourse.group/t/non-null-assertion-operator-aka-swifts-force-unwrap/1085)

**Comment by hax** [comment]
I strongly prefer do not include `!.` in this proposal so we can have separate proposal to add runtime behavior in the future.

**Comment by wparad** [comment]
I also agree that limiting the introduction to some core changes and the expanding it subsequently is best to ensure the introductions at each stage make sense. `!.` doesn't have anything to do with types, so arguably it doesn't belong in the proposal let alone having runtime behavior.

**Comment by wparad** [comment]
> The annotation semantics are specific to the tool being used, so the language cannot know their purpose, and adding purpose later would break code.

JS added `?.` without it being a breaking change, so there must be a way to add `!.`, I'm not following the problem you are suggesting would exist.

**Comment by matthewp** [comment]
> JS added ?. without it being a breaking change, so there must be a way to add !., I'm not following the problem you are suggesting would exist.

 `?.` was not valid syntax. If we are saying that `!.` is valid syntax then we cannot add meaning to it later.

1. `x!.foo` is a comment.
  a. Tool A interprets `!` to be a cast.
  b. Tool B interprets `!` to be an assertion that x exists.
2. JavaScript adds meaning to `!`, choosing either Tool A, Tool B, or another Tool X's semantics.
3. Code that expected the other semantics breaks.

**Comment by wparad** [comment]
who is saying `!.` is valid syntax, I'm saying we shouldn't put `!.` in this proposal at all, not usable at development time and not available used at runtime. There wouldn't be a problem then, right?

**Comment by matthewp** [comment]
@wparad I think I misinterpreted your comments to mean to leave it as specified and to then expand later. If you're proposing removing it entirely then I agree you are correct.

**Comment by hax** [comment]
Besides possible runtime behavior, `!.` also have other issues.

One (small but important) issue is, it's actually not `!.` operator but `!` postfix operator, but because of the short-circuiting semantic issue of `a?.b!`, TS actually treat it differently with `a?.b!.c`. So strictly speaking it's two operators: postfix `!` and infix `!.`.

And note there is also another issue: `a!.foo` `a![foo]` `a!(...args)` syntax are not consistent with `a?.foo` `a?.[foo]` `a?.(...args)`.

Due to these issues (maybe there are others I missed), I believe we'd better split it to a separate proposal.

**Comment by Jamesernator** [comment]
From my experience both working in a company previously that used TypeScript and from my own usage having `!` have runtime semantics would be extremely useful.

In my previous job the assertion was used fairly extensively, however I also noticed in it's usage that such assertions were BY FAR more likely to be incorrect compared to other type assertions. This would lead to runtime exceptions or silent errors happening further down, often far from the point of assertion because such assertions were often used on parameters and things.

A lot of the reason `!` winds up incorrect, is because it very often the case that something should be nullable except when in certain states. While TypeScript and such do provide discrimated unions and such, they can't quite cover all patterns and possible states where *actually that value can be null* could slip through.

Having `!` would incredibly help detect to silent errors and mistakes, and even runtime errors that cause "Cannot read properties of undefined" would be pushed all the way back to the actual source of the error, not somewhere at a far distance.
